File: tests/Billing.Tests/ValidationRuleConfigTests.cs
using Billing.Config.TenantConfigSync.Internal;
using Billing.Config.TenantConfigSync.Internal.Dto;
using Billing.Mediation.Validation;
using MediationModel;

namespace Billing.Tests;

/// <summary>The data-driven config path that replaces the legacy MEF/Spring.NET rule composition: validation
/// rules are BEHAVIOUR in a name-keyed registry, and config-manager sends only rule REFERENCES (name + data).
/// The mapper binds them, fail-fast on an unknown name.</summary>
public class ValidationRuleConfigTests
{
    [Fact]
    public void Registry_resolves_known_rules_and_fails_fast_on_unknown()
    {
        var rule = ValidationRuleRegistry.Default.Resolve("InPartnerIdGt0");
        Assert.True(rule.Validate(new cdr { InPartnerId = 5 }));
        Assert.False(rule.Validate(new cdr { InPartnerId = 0 }));

        var threshold = ValidationRuleRegistry.Default.Resolve("InPartnerCostGt", 1.0m);   // data-parameterised
        Assert.True(threshold.Validate(new cdr { InPartnerCost = 2.0m }));
        Assert.False(threshold.Validate(new cdr { InPartnerCost = 0.5m }));

        Assert.Throws<InvalidOperationException>(() => ValidationRuleRegistry.Default.Resolve("NopeRule"));
    }

    [Fact]
    public void Checklists_load_from_config_as_rule_references()
    {
        var dto = new TenantDto
        {
            Name = "admin", DbName = "telcobright",
            Context = new DynamicContextDto
            {
                MediationContext = new MediationContextDto
                {
                    ServiceGroupConfigurations = new Dictionary<int, ServiceGroupConfigDto>
                    {
                        [10] = new()
                        {
                            ServiceGroupId = 10,
                            Rules = new List<RatingRuleDto> { new() { IdServiceFamily = 10, AssignDirection = 1 } },
                            AnsweredChecklist = new List<RuleRefDto>
                            {
                                new() { Rule = "InPartnerIdGt0" },
                                new() { Rule = "InPartnerCostGt", Data = 0.0m },
                            },
                        },
                    },
                    CommonChecklist = new List<RuleRefDto> { new() { Rule = "ServiceGroupGt0" } },
                },
            },
        };

        var med = ConfigManagerMapper.ToTenant(dto).Context.MediationContext;

        Assert.Single(med.CommonChecklist);                            // common checklist bound from the ref
        var sg10 = med.ServiceGroupConfigurations[10];
        Assert.Single(sg10.Rules);                                     // rating-rule data survived
        Assert.Equal(2, sg10.AnsweredChecklist.Count);                 // two refs -> two real rules
        Assert.Empty(sg10.UnansweredChecklist);

        // and the resolved rules actually validate
        Assert.False(sg10.AnsweredChecklist[0].Validate(new cdr { InPartnerId = 0 }));
        Assert.True(sg10.AnsweredChecklist[0].Validate(new cdr { InPartnerId = 5 }));
    }

    [Fact]
    public void Unknown_rule_in_config_fails_the_load()
    {
        var dto = new TenantDto
        {
            Name = "admin", DbName = "telcobright",
            Context = new DynamicContextDto
            {
                MediationContext = new MediationContextDto
                {
                    CommonChecklist = new List<RuleRefDto> { new() { Rule = "DoesNotExist" } },
                },
            },
        };

        Assert.Throws<InvalidOperationException>(() => ConfigManagerMapper.ToTenant(dto));
    }
}

File: java/src/test/java/com/telcobright/billing/tenantconfigsync/internal/ValidationRuleConfigTests.java
package com.telcobright.billing.tenantconfigsync.internal;

import com.telcobright.billing.mediation.engine.models.cdr;
import com.telcobright.billing.mediation.validation.ValidationRuleRegistry;
import com.telcobright.billing.tenantconfigsync.internal.dto.DynamicContextDto;
import com.telcobright.billing.tenantconfigsync.internal.dto.MediationContextDto;
import com.telcobright.billing.tenantconfigsync.internal.dto.RatingRuleDto;
import com.telcobright.billing.tenantconfigsync.internal.dto.RuleRefDto;
import com.telcobright.billing.tenantconfigsync.internal.dto.ServiceGroupConfigDto;
import com.telcobright.billing.tenantconfigsync.internal.dto.TenantDto;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * The data-driven config path that replaces the legacy MEF/Spring.NET rule composition: validation rules are
 * BEHAVIOUR in a name-keyed registry, and config-manager sends only rule REFERENCES (name + data). The mapper
 * binds them, fail-fast on an unknown name. Faithful port of ValidationRuleConfigTests.cs.
 *
 * <p>PACKAGE NOTE: the C# test lived in {@code Billing.Tests} and reached {@code internal ConfigManagerMapper}
 * via InternalsVisibleTo. The Java {@code ConfigManagerMapper} is package-private in
 * {@code tenantconfigsync.internal}, so per RULE T0 (same package for package-private access) this test is
 * placed HERE rather than in {@code mediation.validation} — that is the only package from which both
 * {@code ConfigManagerMapper} and the public {@code ValidationRuleRegistry} are reachable. C#
 * {@code InvalidOperationException} maps to {@link IllegalStateException}; {@code 1.0m}/{@code 0.0m} to
 * {@link BigDecimal} strings; record accessors take {@code ()}.</p>
 */
class ValidationRuleConfigTests {

    @Test
    void Registry_resolves_known_rules_and_fails_fast_on_unknown() {
        var rule = ValidationRuleRegistry.Default.Resolve("InPartnerIdGt0");
        cdr ok = new cdr();
        ok.InPartnerId = 5;
        assertTrue(rule.Validate(ok));
        cdr zero = new cdr();
        zero.InPartnerId = 0;
        assertFalse(rule.Validate(zero));

        // DurationSec-gated: at/above the threshold the field must be > 0; below it, must be 0.
        var gated = ValidationRuleRegistry.Default.Resolve("InPartnerCostGt0", new BigDecimal("1.0"));
        cdr above = new cdr();
        above.DurationSec = BigDecimal.valueOf(60);
        above.InPartnerCost = new BigDecimal("2.0");
        assertTrue(gated.Validate(above));
        cdr aboveZeroCost = new cdr();
        aboveZeroCost.DurationSec = BigDecimal.valueOf(60);
        aboveZeroCost.InPartnerCost = BigDecimal.ZERO;
        assertFalse(gated.Validate(aboveZeroCost));
        cdr below = new cdr();
        below.DurationSec = BigDecimal.ZERO;
        below.InPartnerCost = BigDecimal.ZERO;
        assertTrue(gated.Validate(below));   // below threshold -> must be 0

        assertThrows(IllegalStateException.class, () -> ValidationRuleRegistry.Default.Resolve("NopeRule"));
    }

    @Test
    void Checklists_load_from_config_as_rule_references() {
        TenantDto dto = new TenantDto();
        dto.Name = "admin";
        dto.DbName = "telcobright";

        DynamicContextDto context = new DynamicContextDto();
        MediationContextDto mediationDto = new MediationContextDto();

        ServiceGroupConfigDto sg = new ServiceGroupConfigDto();
        sg.ServiceGroupId = 10;
        RatingRuleDto ratingRule = new RatingRuleDto();
        ratingRule.IdServiceFamily = 10;
        ratingRule.AssignDirection = 1;
        sg.Rules = new ArrayList<>(List.of(ratingRule));
        RuleRefDto ans1 = new RuleRefDto();
        ans1.Rule = "InPartnerIdGt0";
        RuleRefDto ans2 = new RuleRefDto();
        ans2.Rule = "InPartnerCostGt0";
        ans2.Data = new BigDecimal("0.0");
        sg.AnsweredChecklist = new ArrayList<>(List.of(ans1, ans2));
        Map<Integer, ServiceGroupConfigDto> sgConfigs = new HashMap<>();
        sgConfigs.put(10, sg);
        mediationDto.ServiceGroupConfigurations = sgConfigs;

        RuleRefDto common = new RuleRefDto();
        common.Rule = "ServiceGroupGt0";
        mediationDto.CommonChecklist = new ArrayList<>(List.of(common));

        context.MediationContext = mediationDto;
        dto.Context = context;

        var med = ConfigManagerMapper.ToTenant(dto).Context.MediationContext;

        assertEquals(1, med.CommonChecklist.size());                            // common checklist bound from the ref
        var sg10 = med.ServiceGroupConfigurations.get(10);
        assertEquals(1, sg10.Rules().size());                                  // rating-rule data survived
        assertEquals(2, sg10.AnsweredChecklist().size());                      // two refs -> two real rules
        assertTrue(sg10.UnansweredChecklist().isEmpty());

        // and the resolved rules actually validate
        cdr zero = new cdr();
        zero.InPartnerId = 0;
        assertFalse(sg10.AnsweredChecklist().get(0).Validate(zero));
        cdr ok = new cdr();
        ok.InPartnerId = 5;
        assertTrue(sg10.AnsweredChecklist().get(0).Validate(ok));
    }

    @Test
    void Unknown_rule_in_config_fails_the_load() {
        TenantDto dto = new TenantDto();
        dto.Name = "admin";
        dto.DbName = "telcobright";

        DynamicContextDto context = new DynamicContextDto();
        MediationContextDto mediationDto = new MediationContextDto();
        RuleRefDto badRef = new RuleRefDto();
        badRef.Rule = "DoesNotExist";
        mediationDto.CommonChecklist = new ArrayList<>(List.of(badRef));
        context.MediationContext = mediationDto;
        dto.Context = context;

        assertThrows(IllegalStateException.class, () -> ConfigManagerMapper.ToTenant(dto));
    }
}
